docs/content/en/functions/transform.Unmarshal.md

```diff
@@ -45,3 +45,23 @@ Example:
 ```go-html-template
 {{ $csv := "a;b;c" | transform.Unmarshal (dict "delimiter" ";") }}
 ```
+
+## XML data
+
+{{% warning "Order of XML elements" %}}
+The order of XML elements and data structures is not preserved. An XML element is an order-dependent collection of anonymous values, while a data structure is an order-independent collection of named values.
+{{% /warning %}}
```

Comment:
I'm not totally sure I follow this logic.

Given the example below, do I not get an ordered slice of items?

```xml
<channel>
			<title>Example feed</title>
			<item>
				<title>Item 1</title>
			</item>
	     <item>
				<title>Item 2</title>
			</item>
		</channel>
```

Reply:
I added this note following the comment of @jmooring https://github.com/gohugoio/hugo/pull/9044#issuecomment-947645742 and my reply below that.

> @vanbroup
> 
> > That limitation doesn't seem to be relevant to the usage in Hugo, does it?
> 
> For ~the majority of~ some use cases I don't think it is relevant. But technically, a XML node has a name, value, and position. When you convert to a map you lose the position.
> 
> data/foo.xml
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <item>
>   <b>This is the first node</b>
>   <a>This is the second node</a>
>   <c>This is the third node</c>
> </item>
> ```
> 
> template
> 
> ```
> <pre>{{ jsonify (dict "indent" "  ") site.Data.foo }}</pre>
> ```
> 
> output
> 
> ```
> {
>   "item": {
>     "a": "This is the second node",
>     "b": "This is the first node",
>     "c": "This is the third node"
>   }
> }
> ```

Reply:
Sure, I understand this re what I consider to be _map values_, but what about this example:

```xml
<channel>
<title>Example feed</title>
<item>
<title>Item 1</title>
</item>
<item>
<title>Item 2</title>
</item>
</channel>
```

Reply:
Correct, I tested to be sure, but the order of a slice is preserved:

```json 
{
  "channel": {
    "item": [
      {
        "title": "Item 1"
      },
      {
        "title": "Item 2"
      }
    ],
    "title": "Example feed"
  }
}
```
But the order of the values in an item not:
```xml
<channel>
    <title>Example feed</title>
    <item>
        <title>Item 1</title>
        <description>Description 1</description>
    </item>
    <item>
        <title>Item 2</title>
        <description>Description 2</description>
    </item>
</channel>
```
Results in a swapped order of the title and description:
```json
{
  "channel": {
    "item": [
      {
        "description": "Description 1",
        "title": "Item 1"
      },
      {
        "description": "Description 2",
        "title": "Item 2"
      }
    ],
    "title": "Example feed"
  }
}
```

I suggest we remove the warning as the order of a map is not relevant to how the values are used in Hugo.

Reply:
We should remove the warning (as it confuses more than it helps), but the order is not irellevant:

* When you iterate a map in a Go template, the keys are sorted (I assume ascending order)
* When you marshal a Go map to JSON, the keys are sorted

What I then would expect for XML is, most importantly, that it is consistent.

We have a `Remarshal` func that has a roundtrip test that it would be good if you could add XML to to verify that these formats can be converted to/from.

Reply:
The XML specification requires a root element, parsing without a root element works but not for lists (such a resources in this example).

```xml
<root>
  <resources>
    <params>
      <byline>picasso</byline>
    </params>
    <src>**image-4.png</src>
    <title>The Fourth Image!</title>
  </resources>
  <resources>
    <name>my-cool-image-:counter</name>
    <params>
      <byline>bep</byline>
    </params>
    <src>**.png</src>
    <title>TOML: The Image #:counter</title>
  </resources>
  <title>Test Metadata</title>
</root>
```

`TestRemarshal` fails because it will encode with the unexpected root element:
```
--- FAIL: TestRemarshal (0.01s)
    remarshal_test.go:128: [XML => yaml] Expected 
        resources:
        - params:
            byline: picasso
          src: '**image-4.png'
          title: The Fourth Image!
        - name: my-cool-image-:counter
          params:
            byline: bep
          src: '**.png'
          title: 'TOML: The Image #:counter'
        title: Test Metadata
        
        got
        root:
          resources:
          - params:
              byline: picasso
            src: '**image-4.png'
            title: The Fourth Image!
          - name: my-cool-image-:counter
            params:
              byline: bep
            src: '**.png'
            title: 'TOML: The Image #:counter'
          title: Test Metadata
        
        diff:
        [root:]
    remarshal_test.go:128: [XML => json] Expected 
        {
           "resources": [
              {
                 "params": {
                    "byline": "picasso"
                 },
                 "src": "**image-4.png",
                 "title": "The Fourth Image!"
              },
              {
                 "name": "my-cool-image-:counter",
                 "params": {
                    "byline": "bep"
                 },
                 "src": "**.png",
                 "title": "TOML: The Image #:counter"
              }
           ],
           "title": "Test Metadata"
        }
        
        got
        {
           "root": {
              "resources": [
                 {
                    "params": {
                       "byline": "picasso"
                    },
                    "src": "**image-4.png",
                    "title": "The Fourth Image!"
                 },
                 {
                    "name": "my-cool-image-:counter",
                    "params": {
                       "byline": "bep"
                    },
                    "src": "**.png",
                    "title": "TOML: The Image #:counter"
                 }
              ],
              "title": "Test Metadata"
           }
        }
        
        diff:
        ["root":]
    remarshal_test.go:128: [XML => toml] Expected 
        title = 'Test Metadata'
        		
        [[resources]]
          src = '**image-4.png'
          title = 'The Fourth Image!'
          [resources.params]
            byline = 'picasso'
        
        [[resources]]
          name = 'my-cool-image-:counter'
          src = '**.png'
          title = 'TOML: The Image #:counter'
          [resources.params]
            byline = 'bep'
        
        got
        [root]
          title = 'Test Metadata'
        [[root.resources]]
          src = '**image-4.png'
          title = 'The Fourth Image!'
          [root.resources.params]
            byline = 'picasso'
        
        [[root.resources]]
          name = 'my-cool-image-:counter'
          src = '**.png'
          title = 'TOML: The Image #:counter'
          [root.resources.params]
            byline = 'bep'
        
        
        
        
        diff:
        [[root] [[resources]] [resources.params]]
    remarshal_test.go:128: [XML => TOML] Expected 
        title = 'Test Metadata'
        		
        [[resources]]
          src = '**image-4.png'
          title = 'The Fourth Image!'
          [resources.params]
            byline = 'picasso'
        
        [[resources]]
          name = 'my-cool-image-:counter'
          src = '**.png'
          title = 'TOML: The Image #:counter'
          [resources.params]
            byline = 'bep'
        
        got
        [root]
          title = 'Test Metadata'
        [[root.resources]]
          src = '**image-4.png'
          title = 'The Fourth Image!'
          [root.resources.params]
            byline = 'picasso'
        
        [[root.resources]]
          name = 'my-cool-image-:counter'
          src = '**.png'
          title = 'TOML: The Image #:counter'
          [root.resources.params]
            byline = 'bep'
        
        
        
        
        diff:
        [[[resources]] [root] [resources.params]]
    remarshal_test.go:128: [XML => Toml] Expected 
        title = 'Test Metadata'
        		
        [[resources]]
          src = '**image-4.png'
          title = 'The Fourth Image!'
          [resources.params]
            byline = 'picasso'
        
        [[resources]]
          name = 'my-cool-image-:counter'
          src = '**.png'
          title = 'TOML: The Image #:counter'
          [resources.params]
            byline = 'bep'
        
        got
        [root]
          title = 'Test Metadata'
        [[root.resources]]
          src = '**image-4.png'
          title = 'The Fourth Image!'
          [root.resources.params]
            byline = 'picasso'
        
        [[root.resources]]
          name = 'my-cool-image-:counter'
          src = '**.png'
          title = 'TOML: The Image #:counter'
          [root.resources.params]
            byline = 'bep'
        
        
        
        
        diff:
        [[resources.params] [root] [[resources]]]
    remarshal_test.go:128: [XML =>  TOML ] Expected 
        title = 'Test Metadata'
        		
        [[resources]]
          src = '**image-4.png'
          title = 'The Fourth Image!'
          [resources.params]
            byline = 'picasso'
        
        [[resources]]
          name = 'my-cool-image-:counter'
          src = '**.png'
          title = 'TOML: The Image #:counter'
          [resources.params]
            byline = 'bep'
        
        got
        [root]
          title = 'Test Metadata'
        [[root.resources]]
          src = '**image-4.png'
          title = 'The Fourth Image!'
          [root.resources.params]
            byline = 'picasso'
        
        [[root.resources]]
          name = 'my-cool-image-:counter'
          src = '**.png'
          title = 'TOML: The Image #:counter'
          [root.resources.params]
            byline = 'bep'
        
        
        
        
        diff:
        [[[resources]] [resources.params] [root]]
```

I can 'fix' this by automatically removing any root element in the `metadecorders` which might give some unexpected results or expect these exceptions in the test.

What do you think @bep?

Reply:
Just drop the remarshal thing for now. One question:

* What parts of Hugo do XML work with with this PR?
* I assume `reosurces.Get` and /data; for /data; what is the root key for `mydata.xml` with root `<root>`? `mydata.root.*`?

Reply:
The root key is defined by the xml, so for RSS the key is `rss` as the document starts with `<rss>` and ends with `</rss>`. 

In this example the root key would be `note` and you would use `note.body`:
```
<?xml version="1.0" encoding="UTF-8"?>
<note>
    <to>Tove</to>
    <from>Jani</from>
    <heading>Reminder</heading>
    <body>Don't forget me this weekend!</body>
</note>
```

The current code works for `resources.Get` and the `/data` directory. It doesn't work for configs as Hugo is not expecting a root element. 

Dropping the root element would work as follows:
```
	case XML:
		var xmlRoot xml.Map
		xmlRoot, err = xml.NewMapXml(data)

		var xmlValue map[string]interface{}
		if err == nil {
			xmlRootName, err := xmlRoot.Root()
			if err != nil {
				return toFileError(f, errors.Wrap(err, "failed to unmarshal XML"))
			}
			xmlValue = xmlRoot[xmlRootName].(map[string]interface{})
		}
```

After which you would skip the root element so just `body` in the example above or this for an RSS feed (`{{ range .channel.item }}` instead of `{{ range .rss.channel.item }}`:

```
{{ with resources.Get "https://example.com/rss.xml" | transform.Unmarshal }}
    {{ range .channel.item }}
        <strong>{{ .title | plainify | htmlUnescape }}</strong><br />
        <p>{{ .description | plainify | htmlUnescape }}</p>
        {{ $link := .link | plainify | htmlUnescape }}
        <a href="{{ $link }}">{{ $link }}</a><br />
        <hr>
    {{ end }}
{{ end }}
```

Reply:
I'll bring @jmooring into this discussion. I think we need to get this right, as these kind of things are extremely hard to take back/change.

```
data
└── notes
    ├── note1.xml
    └── note2.xml
```

So, given the above, we currently get `site.Data.notes.note1.note.title` (or something). Which seems a little stuttery to me ... Also, it would be nice to get the different types (JSON, XML, YAML, TOML) as similar as possible in the structure.

Reply:
It would indeed be `site.Data.notes.note1.note.title`

I also have a version ready that would strip the root element which would make it `site.Data.notes.note1.title`. But the root element would still be in the source file which could be confusing.

For frontmatter and configs we could wrap the information in a `<hugo>` root element and only trip the root element if it's called `hugo`, which makes the `TestRemarshal` happy.

Reply:
>For frontmatter and configs

I need to sleep on this, but for frontmatter and config (if we want to got down that route) I would imagine that we would repeat the _topic_ so it would be `<frontmatter></frontmatter>` and `<config></config>` (the latter in a file named `config.xml`).

Reply:
XML for frontmatter and config? Please, no, no, no. 

I don't like the idea of swallowing the root node, because it might provide context. This is contrived, but worth thinking about.

```text
data
└── messages/
    ├── message-1.xml
    ├── message-2.xml
    └── message-3.xml
````

data/messages/message-1.xml

```xml
<info>
  <body>The engine is warm.</body>
</info>
```

data/messages/message-1.xml

```xml
<info>
  <body>The engine is warm.</body>
</info>
```

Reply:
Yeah, well, those aren't messages. If they were messages, they'd be wrapped in a message (root) node. I think it's reasonable to tell a site author, "Hey, if you drop an XML file in the data/books/ directory, it's a book."

So swallow the root node. Sorry for the noise.

Reply:
The reason to swallow the root node would be to support XML for frontmatter and config as these have a defined format and do not expect the root node.

If we don't care about XML for frontmatter and config, which I can't imagine anyone would want to use, we should keep the root node.

The good thing is, if we would want to support XML for frontmatter or config later, we can just swallow these specific root nodes (or in that context somehow).

I would suggest to leave it as it's now, just for `resources.Get` and `/data` without frontmatter and config support, and without modifying the XML (inclusive the root node).

Reply:
> The reason to swallow the root node would be to support XML for frontmatter and config

And to make XML consumption like JSON/YAML/TOML consumption... bep's point about stuttery syntax is dead on.

A site author says, "So let me get this straight. If I want to access the title, I can do `site.Data.books.mybook.title` if it's a JSON/YAML/TOML file, but I have to do `site.Data.books.mybook.book.title` if it's an XML file. Why did they decide to do it that way?" I don't want to have answer that question.

Reply:
Is that not exactly the question to expect when you modify the XML?

XML comes with the root element to my knowledge you always need to specify the root element when querying the contents.

It's simple to remove, but I would be concerned that this would raise questions instead of removing them.

Reply:
> I would be concerned that this would raise questions

Can you provide an example?

Reply:
Simply the reversed, I have XML document file.xml and I want to access the title but this doesn't work:

```
{{ site.Data.directory.file.note.title }}
```

Reply:
OK, take this example:

```
data
└── notes
    ├── note1.xml
    ├── note2.xml
    └── note3.json
```

The above matches pretty much what we have in the front matter world in Hugo (some prefer YAML, others TOML).

If we preserve the XML root, we would need to have something like this:

```
{{ $title := "" }}
{{ $note :=  site.Data.notes.note1 }}
{{ if dataFileIsXML }}
{{ $title = $note.note.title }}
{{ else }}
{{ $title = $note.title }}
{{ end }}
```

Reply:
With a realistic directory structure, that would be more like `{{ site.Data.notes.note_1.note.title }}`. Personally, I'd prefer that XML, JSON, TOML, and YAML all behave the same way. It's a feature:

"As a convenience, Hugo allows you to access XML data in the same way that you access JSON, TOML, and YAML: you do not need to specify the root element."

Reply:
I just checked some XML implementations and they don't seem to be consistent with the root element (e.g., Python vs PHP)

So maybe it's not a problem to remove the root node.

I will push the update to remove the root node and add this to the documentation.

Do we agree that we do not implement and test XML for config and frontmatter?

Reply:
>Do we agree that we do not implement and test XML for config and frontmatter?

Yes.

And also yes to removing the root node for the others. With that I guess the Remarshal tests should pass?

Reply:
No, only if I also add frontmatter support

Reply:
By removing the root node, minimizing xml files will fail as the output is not equal to the input (missing the root node).